js/private/js_devserver.js

```diff
@@ -0,0 +1,161 @@
+const path = require('path')
+const fs = require('fs')
+const os = require('os')
+const child_process = require('child_process')
+
+const RUNFILES_ROOT = path.join(
+    process.env.RUNFILES,
+    process.env.JS_BINARY__WORKSPACE
+)
+
+const synced = new Map()
+
+async function mkdirp(p) {
+    if (p && !fs.existsSync(p)) {
+        await mkdirp(path.dirname(p))
+        await fs.promises.mkdir(p)
+    }
+}
+
+function isExecutable(stats) {
+    return (
+        !!(stats.mode & fs.constants.S_IXUSR) ||
+        !!(stats.mode & fs.constants.S_IXGRP) ||
+        !!(stats.mode & fs.constants.S_IXOTH)
+    )
+}
+
+async function syncRecursive(src, dst) {
+    try {
+        const lstat = await fs.promises.lstat(src)
+        const last = synced.get(src)
+        if (!lstat.isDirectory() && last && lstat.mtimeMs == last) {
+            // this file is already up-to-date
+            return
+        }
+        synced.set(src, lstat.mtimeMs)
+        const exists = fs.existsSync(dst)
+        if (lstat.isSymbolicLink()) {
+            console.log(
+                `Syncing symlink ${src.slice(RUNFILES_ROOT.length + 1)}`
+            )
+            await mkdirp(path.dirname(dst))
+            if (exists) {
+                await fs.promises.unlink(dst)
+            }
+            await fs.promises.symlink(src, dst)
+        } else if (lstat.isDirectory()) {
+            await mkdirp(dst)
+            const contents = await fs.promises.readdir(src)
+            await Promise.all(
+                contents.map((entry) =>
+                    syncRecursive(path.join(src, entry), path.join(dst, entry))
+                )
+            )
+        } else {
+            console.log(`Syncing file ${src.slice(RUNFILES_ROOT.length + 1)}`)
+            await mkdirp(path.dirname(dst))
+            if (exists) {
+                await fs.promises.unlink(dst)
+            }
+            await fs.promises.copyFile(src, dst)
+            const mode = isExecutable(lstat) ? 0o744 : 0x644
+            await fs.promises.chmod(dst, mode)
+        }
+    } catch (e) {
+        console.error(e)
+        process.exit(1)
+    }
+}
+
+async function sync(files, rootdir, sync = false) {
+    for (const file of files) {
+        const src = path.join(RUNFILES_ROOT, file)
+        const dst = path.join(rootdir, file)
+        if (sync) {
+            await syncRecursive(src, dst)
```

Comment:
I think this can be done concurrently for all `files`?

```
const syncedFiles = files.map(file = > syncRecurse(
    path.join(RUNFILES_ROOT, file),
    path.join(rootdir, file)
);

if (sync) {
   await Promise.all(syncedFiles);
}
```
?

Reply:
Fixed up